File: async/javastore/app/controllers/Application.java
package controllers;

import com.fasterxml.jackson.databind.JsonNode;
import models.PaymentService;
import models.StockQuery;
import play.data.Form;
import play.libs.F;
import play.mvc.Controller;
import play.mvc.Result;
import views.html.index;

import java.util.List;

import static play.libs.F.Promise;

public class Application extends Controller {

    private static final StockQuery stockQuery = new StockQuery();
    private static final PaymentService paymentService = new PaymentService();

    public static Result index() {
        return ok(index.render());
    }

    public static Promise<Result> search(String query) {
        Promise<JsonNode> resultPromise = stockQuery.searchStock(query);
        return resultPromise.map(new F.Function<JsonNode, Result>() {
            @Override
            public Result apply(JsonNode jsonNode) throws Throwable {
                boolean first = true;
                String result = "Found results: ";
                List<String> descriptions = jsonNode.findValuesAsText("description");
                for (String description : descriptions) {
                    if (!first) {
                        result += ", ";
                    }
                    result += description;
                    first = false;
                }
                result += ".\n";
                return ok(result);
            }
        }).recover(new F.Function<Throwable, Result>() {
            @Override
            public Result apply(Throwable e) throws Throwable {
                return internalServerError(e.getMessage());
            }
        });
    }

    public static class Payment {
        public Integer amount;
    }
    private static Form<Payment> paymentForm = Form.form(Payment.class);

    public static Promise<Result> payments() {
        Payment payment = paymentForm.bindFromRequest().get();
        Promise<String> stringPromise = paymentService.proceedPayments(payment.amount);
        return stringPromise.map(new F.Function<String, Result>() {
            @Override
            public Result apply(String result) throws Throwable {
                return ok(result);
            }
        }).recover(new F.Function<Throwable, Result>() {
            @Override
            public Result apply(Throwable e) throws Throwable {
                return internalServerError(e.getMessage());
            }
        });
    }
  
}

File: async/scalastore/app/controllers/Application.scala
package controllers

import play.api.mvc._
import play.api.data._
import play.api.data.Forms._

import scala.concurrent.ExecutionContext.Implicits.global
import models.{Payment, StockQuery}
import scala.util.{Success, Failure}

object Application extends Controller with StockQuery with Payment {
  
  def index = Action {
    Ok(views.html.index())
  }

  def search(query: String) = Action {
    Async {
      searchStock(query).map {
        case Success(results) => Ok(results.map(_ \ "description").mkString("Found results: ", ", ", ".\n"))
        case Failure(e)       => InternalServerError(e.getMessage)
      }
    }
  }

  val paymentForm = Form("amount" -> number)

  def payments = Action { implicit request =>
    val amount = paymentForm.bindFromRequest.get
    Async {
      proceedPayments(amount).map {
        case Success(msg) => Ok(msg)
        case Failure(e)   => InternalServerError(e.getMessage)
      }
    }
  }
}
